### ISSUE TITLE
InvalidCastException with simple Select() - very specific scenario

### ISSUE THREAD

**Issue by kierenj** [dotnet/efcore#6506]
### Steps to reproduce

This took a very long time to narrow down from a much larger context and codebase... I hope it's worth it :)

https://github.com/RedRiverSoftware/ef-repros/tree/query-build-exception

Example uses sqllocaldb.  git clone, restore, ef database update, dotnet run.
### The issue

Here's the core query:

``` c#
var list = ctx.Set<Widget>()
  .Select(m => new { Manufacturer = m.ManufacturedIdentity.Manufacturer.Name, GeoArea = m.GeoArea.Name })
  .ToList();
```

Removing either field from the select expression makes the issue stop occurring.

I believe the field names are important - I only finally managed to repro when I renamed the entity models/properties to match my real-world case (though still not exacty; GeoArea class is called Region and the prop on Widget for it is called Source).

```
Exception message: System.InvalidCastException: Unable to cast object of type 'System.Linq.Expressions.Expression`1[System.Func`2[Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor+T
ransparentIdentifier`2[Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor+TransparentIdentifier`2[ConsoleApplication.Widget,System.Collections.Generic.IEnumerable`1[ConsoleApplication.
GeoArea]],ConsoleApplication.GeoArea],System.Guid]]' to type 'System.Linq.Expressions.ConstantExpression'. 

Stack trace:
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.QueryFlattener.Flatten(MethodCallExpression methodCallExpression)                                                          
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.OptimizeJoinClause(JoinClause joinClause, QueryModel queryModel, Int32 index, Action baseVisitAction, MethodInfo operatorTo
Flatten, Boolean outerJoin)                                                                                                                                                                      
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitGroupJoinClause(GroupJoinClause groupJoinClause, QueryModel queryModel, Int32 index)                                  
   at Remotion.Linq.QueryModelVisitorBase.VisitBodyClauses(ObservableCollection`1 bodyClauses, QueryModel queryModel)                                                                            
   at Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)                                                                                                                 
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitQueryModel(QueryModel queryModel)                                                                                         
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitQueryModel(QueryModel queryModel)                                                                                     
   at Microsoft.EntityFrameworkCore.Query.Internal.SqlServerQueryModelVisitor.VisitQueryModel(QueryModel queryModel)                                                                             
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateQueryExecutor[TResult](QueryModel queryModel)
```
### Further technical details

EF Core version: 1.0.0
Operating system:  Windows 10
Visual Studio version: n/a commandline

**Comment by smitpatel** [comment]
Dupe of #4539

**Comment by kierenj** [comment]
So I kind of grasp your branching strategy, but not quite enough to know if there's something semi stable with a fix for this in.. Can you advise please?

**Comment by kierenj** [comment]
I worked it out in the end.. added the aspnetcore-dev myget feed and randomly Googled/updated versions across my project tree until everything was resolved.

I do now get another Exception when the same query has no results (in my main project, I haven't tried the repro project) - internal EF one saying a nullable object must have a value.  I'll see if I can repro and open another issue.  (Occurs on .Select() or .Count() so far)
